[FeatureHighlight] Add custom fonts

ianegordon opened the pull request:
Exposes a pair of custom font properties.
These properties integrate with the UIAppearance proxy.

Commit:
Automatic changelog preparation for release.

Commit:
Remove references to scripts/release/ paths.

Commit:
Remove excess changelog content.

Commit:
Update CHANGELOG description.

Commit:
[FeatureHighlight] Add custom fonts

Commit:
[FeatureHighlight] Add custom fonts

Commit:
Bump the release.

randallli commented on components/FeatureHighlight/src/MDCFeatureHighlightView.h:
```diff
@@ -20,7 +20,11 @@
 @property(nonatomic, strong, nullable) UIColor *innerHighlightColor UI_APPEARANCE_SELECTOR;
 @property(nonatomic, strong, nullable) UIColor *outerHighlightColor UI_APPEARANCE_SELECTOR;
+
+@property(nonatomic, strong, nullable) UIFont *titleFont UI_APPEARANCE_SELECTOR;
```
unit test?

ianegordon replied:
Added.

randallli replied:
did you forget to add it to the PR?

ianegordon replied:
UI Test added.

If you are specifically talking about a Unit test for titleFont, it is simply wired to a backing _ivar.

Let's catch up offline if there are still open questions.

randallli replied:
oh I was expecting something run by a test. Are you saying you created a manual example to test?

randallli replied:
I would like a unit test that checks that the backing label's font changes when you set the public api.

Since we are embracing a "I will set a different font then they one you want" (dynamic text) we should unit test that as well.

Also we should test that the default font gets used when we set the font to nil.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -111,15 +113,15 @@ - (instancetype)initWithFrame:(CGRect)frame {
 
     _titleLabel = [[UILabel alloc] initWithFrame:CGRectZero];
-    _titleLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:MDCFontTextStyleTitle];
+    _titleLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:kTitleTextStyle];
```
this is suppose to query the singleton.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -111,15 +113,15 @@ - (instancetype)initWithFrame:(CGRect)frame {
 
     _bodyLabel = [[UILabel alloc] initWithFrame:CGRectZero];
-    _bodyLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:MDCFontTextStyleSubheadline];
+    _bodyLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:kBodyTextStyle];
```
suppose to use the singleton

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -178,12 +180,70 @@ - (void)setOuterHighlightColor:(UIColor *)outerHighlightColor {
+    }
+  } else {
+    if (_mdc_adjustsFontForContentSizeCategory) {
+      _bodyLabel.font = [UIFont mdc_preferredFontForMaterialTextStyle:kBodyTextStyle];
```
we want to use the singleton's value here too

ianegordon replied:
Done.

randallli replied:
This does not look like it changed.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -178,12 +180,70 @@ - (void)setOuterHighlightColor:(UIColor *)outerHighlightColor {
+    if (_mdc_adjustsFontForContentSizeCategory) {
+      _bodyLabel.font = [UIFont mdc_preferredFontForMaterialTextStyle:kBodyTextStyle];
+    } else {
+      _bodyLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:kBodyTextStyle];
```
singleton's value.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -178,12 +180,70 @@ - (void)setOuterHighlightColor:(UIColor *)outerHighlightColor {
   _titleLabel.textColor = titleColor;
 }
 
+- (void)setBodyFont:(UIFont *)bodyFont {
```
I think this should read:
```
- (void)setBodyFont:(UIFont *)bodyFont {
  if (bodyFont == nil) {
    _bodyLabel.font = [Typography bodyTextFont];
  } else {
    _bodyLabel.font = bodyFont;
  }
  
  if (_mdc_adjusts...) {
    _bodyLabel.font =
        [_bodyLabel.font mdc_fontSizedForMaterialTextStyle:kBodyTextStyle
                                                scaledForDynamicType:_mdc_adjustsFontForContentSizeCategory];
  }
}
```

randallli replied:
and that changes to the _mdc_adjust bool and sizeCategoryChanges should just call `[self setBodyFont:self.bodyFont]`

ianegordon replied:
fontSizedForMaterialTextStyle is intended to scale a custom font.
preferredFontForStyle is intended to eventually migrate to UIContentSizeAdjusting, which is Apple's prescribed API.

Per the Design Doc, properties (set*Font and mdc_setAdjusting) and the sizeCategoryChange handler (and any other font changing method) should call update*Font.

To be clear, this is just a question of whether the code is part of the setter or encapsulated in its own method.  Due to possible KVO side-effects I recommend against implicitly calling complex setters from other methods.

randallli replied:
Sorry i had a minor error in my suggestion. I corrected it (inside the if (_mdc_adjusts...){ clause i changed the object calling of the mdc_fontSizedForMaterial.. method)

If we want the updateFont to be in a different method fine but it should source from the label.font not the property that we have created.

ianegordon replied:
Unfortunately, it won't be orthogonal.

setFont - Comic@22
enable Dynamic Type auto-adjusting : Comic@22 gets scaled to Comic@26
disable Dynamic Type auto-adjusting :
We need the original Comic@22, but we can't get back to there from Comic@26

We need to store the customFont as explicit state.

randallli replied:
Yes we talked about that and we decided that it was acceptable to lose size. We would use the "large" or what apple considers default when you turned off the dynamic font.
Since this is developer API action they have a workaround to manually set the precise size again if they need to.

jverkoey replied:
@randallli it sounds like there might be different assumptions in this discussion. I think what Ian is doing is in line with what we discussed last, so I'll clarify my primary assumption here to make sure it matches yours.

If someone sets a font, it's important that enabling/disabling dynamic type (using the mdc_adjusts bool) does not change the font property.

```swift
// In unit test form:
featureHighlight.font = SomeCustomFont
featureHighlight.mdc_adjusts = true
featureHighlight.mdc_adjusts = false
XCTAssertEqual(featureHighlight.font, SomeCustomFont)
```

It sounds like you feel like the unit test should look like so:

```swift
// In unit test form:
featureHighlight.font = SomeCustomFont
featureHighlight.mdc_adjusts = true
featureHighlight.mdc_adjusts = false
XCTAssertNotEqual(featureHighlight.font, SomeCustomFont)
```

Which I believe we discussed could be confusing to a client (side effects should be avoided!).

If you feel strongly about the alternate implementation @randallli it may be in our best interest to land this change for now and then explore the merits of the alternative approach in a follow-up so that we're not blocking the general body of work here.

randallli replied:
I expect this unit test to pass:

Typogragrphy.fontloader = ComicSans
featureHighlight.mdc_adjust = true
XCTAssert(featureHighlight.font.face, comicSans)
featureHighlight.mdc_adjust = false
XCTAssert(featureHighlight.font.face, comicSans)

With the current logic both fail.

Commit:
[ButtonBar, NavigationBar] Add BUILD file for kokoro (#2598)

* [ButtonBar, NavigationBar] Add BUILD file for kokoro

ButtonBar ColorThemer depends on NavigationBar
NavigationBar depends on ButtonBar

Commit:
Utilize MDCTypography if custom. Add sample.

Commit:
Utilize MDCTypography if custom. Add sample.

ianegordon commented:
This component never used the MDCTypography singleton, and I wasn't intending to add support.
In retrospect, having an easy transition for people who are using the singleton will likely make our lives easier.  MDCTypography added where appropriate.

Commit:
Remove extra method

Commit:
Remove extra method

ianegordon commented:
PTAL.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -111,15 +113,29 @@ - (instancetype)initWithFrame:(CGRect)frame {
-    _titleLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:MDCFontTextStyleTitle];
+    // If we are using the default (system) font loader, retrieve the
+    // font from the UIFont standardFont API.
+    if ([MDCTypography.fontLoader isKindOfClass:[MDCSystemFontLoader class]]) {
```
This is not correct anymore since we are going to add a fontloader for each style. Or we are creating a delegate. This is how we globally set a default font faces for all our styles.

The MDCSystemFontLoader should already return the [UIFont mdc_standardFontForMaterialTextStyle:kTitleTextStyle]

concretely we should drop the if clause and just put:
`_titleLabel.font = [MDCTypography titleFont];`

randallli replied:
by copying this logic that will be in [MDCTypography titleFont] you are making it harder for us to pivot if we determine that UIAppearance is not scalable for our needs in GMDC

randallli replied:
by accessing MDCTypography.fontloader you are increasing usage of the very thing you are trying to delete.

ianegordon replied:
Do we have evidence the UIAppearance is not scalable?
We already evaluated UIAppearance and decided it was the path forward when we added the color themers.  Happy to revisit that decision, but it might be out of scope for this PR.

As far as utility:

If I want all of my buttons to use GoogleSans:
[MDCButton appearance].titleFont = [UIFont fontNamed:@"Google-Sans"];

Following the pattern of our color themers if we want to turn use Roboto everywhere
MDCFontScheme *scheme = [MDCFontScheme schemeWithFontFamily:@"Roboto"];
[MDCAppBarFontThemer apply:scheme to: [MDCAppBarView appearance]];
[MDCButtonFontThemer apply:scheme to:[MDCButton appearance]];
[MDCFeatureHighlightThemer apply:scheme to:[MDCFeatureHighlightThemer appearance]];
// etc.

Note that this would be the migration path for people using the RobotoFontLoader.

randallli reviewed (dismissed):
please dismiss my review when you want me to take another look

Commit:
Merge branch 'release-candidate' into develop

Commit:
Fix minor typos in the release publish command.

Commit:
Automatic changelog preparation for release.

Commit:
Clean up the changelog.

willlarche committed:
[TextFields] Feature branch merge: Clarity, Safety, Docs, Comments, Formatting (#2634)

* [TextFields] Full width changes. (#2590)

* [TextFields] Propagating changes to themers. (#2585)

* [TextFields] Propagating changes to themers.

* [TextFields] Updates to themer examples.

* [TextFields] Base controller class files (#2582)

* [TextFields] New Base class files.

* [TextFields] Naming corrections.

* [TextFields] Formatting.

* [TextFields] Replacing -Default with -Underline (#2583)

* [TextFields] Adding underline controller.

* [TextFields] Changes to umbrella import file and removing default controllers.

* [TextFields] Renaming one file.

* [TextFields] Updated controller headers. (#2584)

* [TextFields] Updating examples. (#2586)

* [TextFields] Running clang-format on the elements. (#2587)

* [TextFields] Running format on the element itself.

* [TextFields] Minor formatting.

* [TextFields] Updating tests and formatting. (#2588)

* [TextFields] Filled controller changes. (#2589)

* [TextFields] Legacy controller comments, naming, and formatting. (#2591)

* [TextFields] Legacy controller comments, naming, and formatting.

* [TextFields] Adding safety to legacy default init.

* [TextFields] Adding safety to Legacy Default init.

* [TextFields] Formatting Legacy Default.

* [TextFields] Formatting Legacy Full Width.

* [TextFields] Text area comments, updated super, naming, formatting. (#2593)

* [TextFields] Text area updates.

* [TextFields] Clarifying comment.

* [TextFields] Outline changes and formatting. (#2592)

* [TextFields] Updates to styling doc. (#2594)

* [TextFields] Updates to styling doc.

* [TextFields] Updating doc.

* [TextFields] Full width comments, safety, formatting, and underline support. (#2633)

* [TextFields] Adding safety to full width.

* [TextFields] Formatting Full Width.

* [TextFields] PR feedback.

* [TextFields] Safety in init of Base (#2632)

* [TextFields] Adding safety to base init.

* [TextFields] Formatting Base.

* [TextFields] Updates to controller header. (#2636)

* [TextFields] Updates to controller header.

* [TextFields] Updating misleading comments.

* [TextFields] Supplemental file name changes. (#2635)

Commit:
Bump the release number to 42.2.0.

Commit:
Merge branch 'release-candidate' into develop

Commit:
Automatic changelog preparation for release.

Commit:
Bump the release.

Commit:
Update changelog.

Commit:
More release notes.

randallli reviewed (changes requested):
I think we need changes to MDCTypography to land before this fully works.

The font loader setters per styles or a delegate.

ianegordon commented:
Happy to discuss offline #2689.

Even with the changes you're proposing, we will have the same interface via the customFont properties.  Are we clear to land this (and the other) components based on the API and we can revisit the implementation after Friday?

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -111,15 +113,29 @@ - (instancetype)initWithFrame:(CGRect)frame {
+      _titleLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:kTitleTextStyle];
+    } else {
+      // There is a custom font loader, retrieve the font from it.
+      _titleLabel.font = [MDCTypography titleFont];
```
The logic inside MDCTypography titleFont will be a copy this if else clause.

If as you propose the source of truth should be mdc_standardFontForMaterialTextStyle... The Typography component should call [UIFont mdc_standardFontForMaterialTextStyle:kTitleTextStyle];

Commit:
Clarify which TextField APIs were renamed.

Commit:
Revert "Clarify which TextField APIs were renamed."

This reverts commit 37f1c9d2b0e612843c98fe5d95e512d99afc1a08.

Commit:
Better migration docs.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -111,15 +113,29 @@ - (instancetype)initWithFrame:(CGRect)frame {
-    _bodyLabel.font = [UIFont mdc_standardFontForMaterialTextStyle:MDCFontTextStyleSubheadline];
+    // If we are using the default (system) font loader, retrieve the
+    // font from the UIFont standardFont API.
+    if ([MDCTypography.fontLoader isKindOfClass:[MDCSystemFontLoader class]]) {
```
The MDCSystemFontLoader class is also kind of an internal implementation detail of MDCTypography.

randallli commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -111,15 +113,29 @@ - (instancetype)initWithFrame:(CGRect)frame {
+      // There is a custom font loader, retrieve the font from it.
+      _bodyLabel.font = [MDCTypography body1Font];
+    }
     _bodyLabel.shadowColor = nil;
```
Just discussed with @jverkoey,
we should add this:
_bodyFont = _bodyLabel.font;

randallli reviewed (dismissed):
i have to add a checklist in a couple moments.

Commit:
[OverlayObserver] Unregister in dealloc (#2686)

When being deallocated, the OverlayObserver should unregister notifications.
On iOS versions before 9 (8 and below), the NSNotificationCenter unsafely
retains references to any observers. On iOS 9 and above, it will keep safe
unretained references.  This bug wasn't found previously because only the
static singleton was ever used by client components and there were no unit
tests for the component.

Closes #2685

Commit:
[Overlay] Add missing QuartzCore dependency (#2692)

Commit:
[HeaderStackView] Add BUILD file and no-op test (#2695)

Commit:
[Slider] Add BUILD file (#2694)

Also fix import in the component and an implicit cast in the tests.

Commit:
[CollectionCells] Add BUILD file (#2693)

Fixing incorrect import in a unit test as well

jverkoey committed:
[docs] Updated with new release process (#2504)

Commit:
Merge branch 'release-candidate' into develop

jverkoey committed:
Delete config.yml

jverkoey committed:
[Tabs] Re-enable new ink (reverts #2665) (#2682)

This reverts commit 83994481a936a47d4bec4381fa34f711c47ba486.

samnm committed:
[Shapes] Custom properties for consistent behavior (#2691)

* [Shapes] Custom properties for consistent behavior

* Ordering

Commit:
[FeatureHighlight] Add custom font tests

Commit:
Merge branch 'fhfont' of github.com:ianegordon/material-components-ios into fhfont

ianegordon commented:
To be reopened in a new PR.
